### DISCUSSION TITLE
Unquoted characters

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#37] [General]
There is a trade-off between disallowing unquoted characters and introducing new functionalities.

**Comment by hchiba1** [comment]
`,` is disallowed for introducing the list of property values.

**Reply by hchiba1** [reply]
No problem in Blitzboard? @matsusho070 @ryotayamanaka

**Reply by matsusho070** [reply]
As of now, since we don't provide sample data that includes `,`, I think there are no issues. (Practically, it would be good to have clear syntax highlighting and error messages in case `,` is mistakenly included.)

**Comment by hchiba1** [comment]
If `:` is disallowed, then URI should be quoted in property values.

**Reply by hchiba1** [reply]
Is there any other reason for Blitzboard to allow `:` in property values? @matsusho070 @ryotayamanaka

**Reply by matsusho070** [reply]
No, and currently, to avoid confusion, strings containing `:` are automatically quoted  during import in Blitzboard sandbox.

**Reply by nichtich** [reply]
Sorry for confusion: what to you refer to with "Blitzboard sandbox"?

Quotation rules on export to PG format should be more strict than parsing rules for hand-written PG format.

**Reply by matsusho070** [reply]
[Here](https://www.blitzboard.io/ ) we have a tool named Blitzboard and "Blitzboard sandbox" is a web site for trial of its functionality.
We use PG format to describe graph data in the sandbox. It also supports PG-JSON as a format to import.

**Comment by hchiba1** [comment]
If `#` is disallowed, then the space is not required before Trailing Comments.

**Comment by nichtich** [comment]
There are four places for unquoted strings, each may have its own disallowed charatcers.

- node identifiers
- labels
- property keys
- property values

Furthermore it makes a difference whether special characters are disallowed anywhere or only at start and/or end of an unquoted string. Disallowed characters are *only* required when whitespace before/after an element is optional!

The current grammar subset use in pgraphs 0.3.1 (without quoted strings and additional rules): 

~~~pegjs
// Both node identifiers and labels (the latter preceded by `:` and optional space)
UnquotedIdentifier = PlainStart PlainChar* // used for both node identifier and label
PlainChar = [^\x20\x09\x0A\x0D"<>{}|^]
PlainStart = ![:(#] PlainChar
// Key is same if followed by space, otherwise everything up to last `:`
Key = UnquotedIdentifier Space ":"  / PlainStart ( ( !":" PlainChar )* ":" )+
// Value is same but must not include comma
Value = PlainStart ( !"," PlainChar )* )
~~~

Reasons for current choices:

- allow as much as possible to be unquoted, in particular any valid URI
- keep rules for identifiers, labels, property keys and values as same as possible
- disallow `:` as first character to avoid confusing identifier with label (can be discussed)
- disallow `#` as first character because this would be a comment when it's the first node. Still it could be allowed as first char in labels and properties but disallwed for conistency to keep rules simpler (can be discussed)
- disallow `(` as first characters to be open for extension and `(a)` may be misread for users of other languages (can be discussed)
- disallow `<>{}|^` because these are not allowed in URIs anyway and to
  -  avoid confusion with direction (`<` and `>` so `a->b` is an edge)
  - support additional line separator (`|`, so `a|b` are two nodes) 
  - allow extension (`{`, `}`, `^`) for hypergraphs, hierchical graphs, datatypes... (can be discussed)
- special rule for `:` in property keys to allow omit space between key, value, and colon

The only missing special rule for exclusion is maybe forbid `--` in node identifiers to allow `a--b` be an edge (but adding a single space would solve this: `a --b`).

**Comment by hchiba1** [comment]
I forgot to add the one:

If `-` is disallowed, `a->b` can be interpreted as edge.

**Reply by nichtich** [reply]
This is also possible if `-` is allowed by lookahead parsing. I've modified pgraph with version 0.3.2 to support condense edges such as `a->b` and `a--b` while still allowing `-` in unquoted identifers. Additional rules compared to my comment above:

- disallow `-` as first character 
- disallow the sequence `--` in an unquoted identifier (I would like to not make this rule because URIs may contain this sequence but it's a trade-of between requirement "allow to write unquoted URIs" and requirement "allow to write condense undirected edges".

**Reply by hchiba1** [reply]
For the meanwhile, maybe we should give up on the idea of condense edges.
